[NOSQUASH] Use unique_ptrs in various places

Desour opened the pull request:
**(*) Squash instructions:** Please fixup all commits that begin with a lowercase letter to the commit that came before.

This is a maintenance/refactor PR.

Clear ownership semantics make understanding code easier, and they can prevent bugs. Smart ptrs make such semantics obvious.
This PR aims to gradually replace some uses of raw ptrs with smart ptrs. I won't replace all at once, but we need to start somewhere.

## To do

This PR is a Ready for Review.

## How to test

* Compile.
* Read through the code.
* Review commits separately!

numberZero commented on src/serverenvironment.cpp:
```diff
@@ -1210,9 +1211,6 @@ void ServerEnvironment::clearObjects(ClearObjectsMode mode)
 
 		// Delete active object
-		if (obj->environmentDeletes())
-			delete obj;
```
The comment now looks obsolete

Desour replied:
It's for the `return`. See comment in server/activeobjectmgr.h.

numberZero commented on src/serverenvironment.cpp:
```diff
@@ -620,7 +620,8 @@ void ServerEnvironment::savePlayer(RemotePlayer *player)
 {
-	PlayerSAO *playersao = new PlayerSAO(this, player, peer_id, is_singleplayer);
+	auto playersao_u = std::make_unique<PlayerSAO>(this, player, peer_id, is_singleplayer);
+	auto playersao = playersao_u.get();
```
I’d move this `get()` right above `addActiveObject` to keep the unique pointer unique for as long as feasible.

Desour replied:
The raw ptr is not meant to be owning, so the unique ptr is always uniquely owned.
Anyway, will do.

numberZero commented on src/server/activeobjectmgr.cpp:
```diff
@@ -25,11 +25,13 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 
-void ActiveObjectMgr::clear(const std::function<bool(ServerActiveObject *, u16)> &cb)
+
+void ActiveObjectMgr::clear(const std::function<bool(ServerActiveObject *, u16)> &cb) //TODO
```
TODO what?

Desour replied:
Oops.

numberZero commented on src/client/clientenvironment.cpp:
```diff
@@ -361,26 +361,27 @@ GenericCAO* ClientEnvironment::getGenericCAO(u16 id)
+	auto obj = object.get();
 	// Register object. If failed return zero id
-	if (!m_ao_manager.registerObject(object))
+	if (!m_ao_manager.registerObject(std::move(object)))
```
Using `obj` after moving `object` doesn’t feel safe. I know it is here and now, but... Murphy awaits.
And I see this pattern all over the ActiveObject commit.

Desour replied:
The semantics are the same as before. So this feeling of unsafety is a good thing, it makes one more wary.

The obj should still be alive if the returned id is not 0, I guess. This is checked everywhere (apart from some places in the unittests).

numberZero reviewed (commented):
Looks good, except of some concerns in the ActiveObject handling.

Desour commented on src/client/activeobjectmgr.cpp:
```diff
@@ -29,9 +29,8 @@ void ActiveObjectMgr::clear()
+		// Note: Object must be marked as gone when children try to detach
+		active_object.second.reset();
 	}
 	m_active_objects.clear();
```
I was worried because of the comment. But as `aomgr::getActiveObject()` also returns null if the obj is no longer in the container, this should be fine. Will do.

numberZero replied:
Oops, you’re right. Reentrancy is scary. Better restore the loop.
Or use something like this:
```suggestion
	// Empty m_active_objects and *then* destroy the objects so their destructors
	// will see an empty map, rather than a map being emptied
	decltype(m_active_objects) objects_to_destroy;
	std::swap(objects_to_destroy, m_active_objects);
```

Desour replied:
I don't think reentrancy of this function matters here. Why would an obj destructor clear the objmgr?

However, reentrance of the container (aka reentrence between the containers member functions) is be an issue. It's probably undefined what `m_active_objects.find()` does while `m_active_objects.clear()` or `m_active_objects.erase()` is executed.
I would've overlooked this, thanks!

Clearing all at once however, is more likely to be wrong. Currently all objects that are alive (and are not the one that is currently being destructed) are in the objmgr. On clear, we kill and remove one obj after another. If we instead take out all at once from the objmgr, and then kill the objs, that's different.

I'll add back the loop with an appropriate comment.

Desour commented on src/client/clientenvironment.cpp:
```diff
@@ -361,26 +361,27 @@ GenericCAO* ClientEnvironment::getGenericCAO(u16 id)
 		ClientActiveObject::create((ActiveObjectType) type, m_client, this);
-	if(obj == NULL)
-	{
+	ClientActiveObject *obj = obj_u.get();
```
Will do.
(FYI, the C++17 thing isn't needed if the declared var is itself the condition to check.)

Desour commented:
Comments addresses. Thanks for the review!

Desour commented:
I've looked in the other changed parts for reentrance issues and found none.
I've also added a comment in activeobjmgr.h about these issues.

----

Sidenote: CAOs call `removeFromScene(true)` in the destructor. But it's also often called before an obj is removed. So, it's essentially called twice. I've left it at that to avoid possible breakage, but a future PR may want to tackle this.

Desour commented:
Rebased.
`detachBlock` (from this commit: <https://github.com/minetest/minetest/commit/ed632f38547e2a63edb1a5dacbc37f3441cc43b6>) now returns the unique_ptr.

Desour commented:
Squashed and rebased.
This now does some improvements on #13468 (had to change the code anyway): The cloned map in master can contain old obj ptrs if some mod removed objects (i.e. via `minetest.clear_objects`). The new code just uses a vector of obj ids (the ids can be outdated, but that doesn't hurt as much). (cc @lhofhansl)

sfan5 commented on src/server/activeobjectmgr.cpp:
```diff
@@ -97,15 +101,17 @@ void ActiveObjectMgr::removeObject(u16 id)
+	// Delete the obj before erasing, as the destructor may indirectly access
+	// m_active_objects.
+	it->second.reset();
+	m_active_objects.erase(it);
```
couldn't `it` also be invalidated here?

Desour replied:
Good point, it can (the destructor of luaentity saos uses the metatable of `core.luaentities`).
Fixed.

Desour commented on src/map.cpp:
```diff
@@ -1892,19 +1891,14 @@ bool ServerMap::deleteBlock(v3s16 blockpos)
-	for (MapBlock *block : m_detached_blocks) {
-		assert(block->isOrphan());
-		delete block;
-	}
```
I had not kept the assert because it caused unused variable warnings in release builds.

Added it back, with a void-cast.

<!--
```cpp
	for (const auto &block : m_detached_blocks) {
		assert(block->isOrphan());
		(void)block; // silence unused-variable warning in release builds
	}

	assert([&]{
		for (const auto &block : m_detached_blocks)
			if (!block->isOrphan())
				return false;
		return true;
	}());

	assert(std::all_of(m_detached_blocks.begin(), m_detached_blocks.end(),
			[](const auto &block) { return block->isOrphan(); }));
```
-->

sfan5 replied:
not sure if void cast forces iteration to happen but this should be fine

Desour commented on src/mapsector.cpp:
```diff
@@ -105,33 +99,39 @@ void MapSector::insertBlock(MapBlock *block)
-	m_blocks.erase(block_y);
+	auto it = m_blocks.find(block_y);
+	assert(it != m_blocks.end());
+	std::unique_ptr<MapBlock> ret = std::move(it->second);
+	assert(ret.get() == block);
+	m_blocks.erase(it);
```
This would free `block`, which we don't want to do in this function.

Desour commented:
Thanks for review! Comments addressed.

I also noticed that `ActiveObjectMgr::step` shouldn't iterate with iterators, as the callback function invokes the `on_step` luaentity callback. Fixed.
(`getObjectsInsideRadius` and `getObjectsInArea` are fine though, as currently we don't pass any `include_obj_cb` that calls into lua.)

sfan5 commented on src/client/activeobjectmgr.cpp:
```diff
@@ -28,11 +28,15 @@ namespace client
-		active_object.second = nullptr;
+		// By resetting to null, the object is also marked as gone, which is
+		// required when children try to detach.
+		active_object.second.reset();
```
the (implicit) iterator could also be invalidated while this runs, in fact
honestly there are way too many pitfalls here.

Desour commented:
I've decided to move the aomgr stuff out into another PR, as it got too complex: #13560

This PR should now be much more trivial.

sfan5 reviewed (approved):
LGTM
